File: jsontocsv.rb
require 'net/http'
require 'json'
require 'csv'

=begin

FILE OVERVIEW:

The idea is to :
1. Get the JSON data
2. Simplify the JSON data (convert if nested)
3. Write it into a csv

=end


=begin

flattenjson(param1, param2):
This is a recursive function which converts a nested json object into a simple json object.
This helps in converting it into a csv file format.

It takes 2 parameters:
#1 param1 => JSON object
#2 param2 => a delimiter to join nested child headings with their parent headings

It returns the input value into a simple JSON object

=end

def flattenjson(data, delim)
    val = {}

    # iterating through input JSON
    data.each do |key, value|

      # if value is a nested JSON, call it again
      if value.class == Hash
        get = flattenjson(value, delim)
        get.each do |key1, value1|

          # This is used to preserve the list data as a string in csv
          if value1.kind_of?(Array)
            value1 = '"'+value1.join(",")+'"'
          end
          val[key+delim+key1] = value1
        end
      else

        # This is used to preserve the list data as a string in csv
        if value.kind_of?(Array)
          value = '"'+value.join(",")+'"'
        end
        val[key] = value
      end
    end
    return val
end

# http GET request on the given URL
url = 'https://gist.githubusercontent.com/romsssss/6b8bc16cfd015e2587ef6b4c5ee0f232/raw/f74728a6ac05875dafb882ae1ec1deaae4d0ed4b/users.json'
uri = URI(url)
response = Net::HTTP.get(uri)
data = JSON.parse(response)

new_json = []
columns = []
for row in data
  # calling the function to reformat json
  # converting original data into easier format to enable easy csv conversion
  value = flattenjson(row, '.')
  columns.push(value.keys)
  new_json.push(value)
end

# getting field names for csv
columns = columns.uniq
CSV.open("ruby_output.csv", "wb") do |csv|

  # writing field names into csv file
  columns.each do |hash|
    csv << hash
  end

  # writing field values
  new_json.each do |hash|
    csv << hash.values
  end

end

File: jsontocsv.py
import requests
import csv


"""
flattenjson(param1, param2):
This is a recursive function which converts a nested json object into a simple json object.
This helps in converting it into a csv file format.

It takes 2 parameters:
#1 param1 => JSON object
#2 param2 => a delimiter to join nested child headings with their parent headings

It returns the input value into a simple JSON object
"""
def flattenjson(data, delim):
  val = {}

  # iterating through input JSON
  for key, value in data.items():

    # if value is a nested JSON, call it again
    if isinstance(value, dict):
      get = flattenjson(value, delim)
      for key1, value1 in get.items():

        # This is used to preserve the list data as a string in csv
        if isinstance(value1, list):
          value1 = '"'+','.join(value1)+'"'

        val[key+delim+key1] = value1
    else:

      # This is used to preserve the list data as a string in csv
      if isinstance(value, list):
        value = '"'+','.join(value)+'"'

      val[key] = value
      
  return val

def main():
  # http GET request on the given URL
  inputUrl = 'https://gist.githubusercontent.com/romsssss/6b8bc16cfd015e2587ef6b4c5ee0f232/raw/f74728a6ac05875dafb882ae1ec1deaae4d0ed4b/users.json'
  r = requests.get(inputUrl)

  # if the response from URL is good
  if r.status_code == 200:
    data = r.json()
    new_json = []

    # converting original data into easier format to enable easy csv conversion
    for row in data:
      value = flattenjson(row,'.')
      new_json.append(value)
      
    # getting field names for csv
    columns = [ x for row in new_json for x in row.keys() ]
    columns = list( set( columns ) )

    # writing into csv file
    with open( "python_output.csv", 'wb' ) as out_file:
      csv_w = csv.writer( out_file )
      csv_w.writerow( columns )

      # writing field values
      for i_r in new_json:
          csv_w.writerow( map( lambda x: i_r.get( x, "" ), columns))
  else:
    # if the JSON request is not good
    print "Could not retrieve JSON from given URL"

if __name__ == "__main__":
    main()
  

  

